src/python_framework_v02/nhd_network.py

```diff
@@ -417,3 +417,85 @@ def replace_waterbodies_connections(connections, waterbodies):
             # copy to new network unchanged
             new_conn[n] = connections[n]
     return new_conn
+
+def build_subnetworks(connections, rconn, max_depth, sources=None):
+    '''
+    Construct subnetworks using a truncated breadth-first-search
+    
+    Arguments:
+        connections
+        rconn
+        max_depth
+        sources
+    Returns:
+        subnetwork_master
+    '''
+    # if no sources provided, use tailwaters
+    if sources is None:
+        # identify tailwaters
+        sources = rconn.keys() - chain.from_iterable(rconn.values())
+    
+    # create a list of all headwaters in the network
+    all_hws = connections.keys() - chain.from_iterable(connections.values())
+
+    subnetwork_master = {}
+    for net in sources:
+
+        # subnetwork creation using a breadth first search restricted by maximum allowable depth
+        new_sources = [net]
+        subnetworks = {}
+        group_order = 0
+        while new_sources:
+
+            # Build dict object containing reachable nodes within max_depth from each source in new_sources
+            rv = {} 
+            for h in new_sources:
+
+                reachable = set()
+                reachable_depth = set()
+                Q = deque([h])
+                D = deque([0])
+                while Q or D:
+
+                    x = Q.popleft() 
+                    y = D.popleft()
+                    reachable.add(x)
+                    reachable_depth.add(y)
+
+                    if len(rconn.get(x, ())) > 1:
+                        us_depth = y+1
+                    else:
+                        us_depth = y
+
+                    if us_depth <= max_depth:
+                        D.extend([us_depth]*len(rconn.get(x, ())))
```

Comment:
Need additional criteria for hitting a minimum number of segments aggregated into the group.

Reply:
I think we have this functionality as it now stands in PR #175 (built from this PR code) So I'll resolve this conversation.